orderfilter/filter.go

```diff
@@ -0,0 +1,210 @@
+	// includes all 0x orders and doesn't add any additional requirements.
+	DefaultCustomOrderSchema = `{}`
+)
+
+var builtInSchemas = []jsonschema.JSONLoader{
+	addressSchemaLoader,
+	wholeNumberSchemaLoader,
+	hexSchemaLoader,
+	orderSchemaLoader,
+	signedOrderSchemaLoader,
+}
+
+func GetDefaultFilter(chainID int) (*Filter, error) {
+	return New(chainID, DefaultCustomOrderSchema)
+}
+
+func GetDefaultTopic(chainID int) (string, error) {
+	defaultFilter, err := GetDefaultFilter(chainID)
+	if err != nil {
+		return "", err
+	}
+	return defaultFilter.Topic(), nil
+}
+
+type Filter struct {
+	topic                string
+	version              int
+	chainID              int
+	rawCustomOrderSchema string
+	orderSchema          *jsonschema.Schema
+	messageSchema        *jsonschema.Schema
+}
+
+func New(chainID int, customOrderSchema string) (*Filter, error) {
+	orderLoader, err := newLoader(chainID, customOrderSchema)
+	if err != nil {
+		return nil, err
+	}
+	rootOrderSchema, err := orderLoader.Compile(rootOrderSchemaLoader)
+	if err != nil {
+		return nil, err
+	}
+
+	messageLoader, err := newLoader(chainID, customOrderSchema)
+	if err := messageLoader.AddSchemas(rootOrderSchemaLoader); err != nil {
+		return nil, err
+	}
+	rootOrderMessageSchema, err := messageLoader.Compile(rootOrderMessageSchemaLoader)
+	if err != nil {
+		return nil, err
+	}
+	return &Filter{
+		chainID:              chainID,
+		rawCustomOrderSchema: customOrderSchema,
+		orderSchema:          rootOrderSchema,
+		messageSchema:        rootOrderMessageSchema,
+	}, nil
+}
+
+func loadExchangeAddress(loader *jsonschema.SchemaLoader, chainID int) error {
+	contractAddresses, err := ethereum.GetContractAddressesForChainID(chainID)
+	if err != nil {
+		return err
+	}
+	// Note that exchangeAddressSchema accepts both checksummed and
+	// non-checksummed (i.e. all lowercase) addresses.
+	exchangeAddressSchema := fmt.Sprintf(`{"enum":[%q,%q]}`, contractAddresses.Exchange.Hex(), strings.ToLower(contractAddresses.Exchange.Hex()))
+	return loader.AddSchema("/exchangeAddress", jsonschema.NewStringLoader(exchangeAddressSchema))
+}
+
+func loadChainID(loader *jsonschema.SchemaLoader, chainID int) error {
+	chainIDSchema := fmt.Sprintf(`{"const":%d}`, chainID)
+	return loader.AddSchema("/chainId", jsonschema.NewStringLoader(chainIDSchema))
+}
+
+func newLoader(chainID int, customOrderSchema string) (*jsonschema.SchemaLoader, error) {
+	loader := jsonschema.NewSchemaLoader()
+	if err := loadChainID(loader, chainID); err != nil {
+		return nil, err
+	}
+	if err := loadExchangeAddress(loader, chainID); err != nil {
+		return nil, err
+	}
+	if err := loader.AddSchemas(builtInSchemas...); err != nil {
+		return nil, err
+	}
+	if err := loader.AddSchema("/customOrder", jsonschema.NewStringLoader(customOrderSchema)); err != nil {
+		return nil, err
+	}
+	return loader, nil
+}
+
+func NewFromTopic(topic string) (*Filter, error) {
+	// TODO(albrow): Use a cache for topic -> filter
```

Comment:
Is this still slated for this PR?

Reply:
No. It's something we should add as part of cross-topic forwarding (or shortly afterwards). Currently we don't call `NewFromTopic` except maybe in the tests. The purpose of this function is to work backwards from a topic string to an order filter.